esrally/client/factory.py

```diff
@@ -238,3 +266,74 @@ def wait_for_rest_layer(es, max_attempts=40):
                 logger.warning("Got unexpected status code [%s] on attempt [%s].", e.status_code, attempt)
                 raise e
     return False
+
+
+def create_api_key(es, client_id, max_attempts=5):
+    """
+    Creates an API key for the provided ``client_id``.
+
+    :param es: Elasticsearch client to use for connecting.
+    :param client_id: ID of the client for which the API key is being created.
+    :param max_attempts: The maximum number of attempts to create the API key.
+    :return: A dict with at least the following keys: ``id``, ``name``, ``api_key``.
+    """
+    logger = logging.getLogger(__name__)
+
+    for attempt in range(1, max_attempts + 1):
+        # pylint: disable=import-outside-toplevel
+        import elasticsearch
+
+        try:
+            return es.security.create_api_key({"name": f"rally-client-{client_id}"})
+        except elasticsearch.TransportError as e:
+            if e.status_code == 405:
+                # We don't retry on 405 since it indicates a misconfigured benchmark candidate and isn't recoverable
+                raise exceptions.SystemSetupError(
+                    "Got status code 405 when attempting to create API keys. Is Elasticsearch Security enabled?", e
+                )
+            else:
+                logger.debug("Got status code [%s] on attempt [%s] of [%s]. Sleeping...", e.status_code, attempt, max_attempts)
+                time.sleep(1)
+
+
+def delete_api_keys(es, ids, max_attempts=5):
+    """
+    Deletes the provided list of API key IDs.
+
+    :param es: Elasticsearch client to use for connecting.
+    :param ids: List of API key IDs to delete.
+    :param max_attempts: The maximum number of attempts to delete the API keys.
+    :return: True iff all provided key IDs were successfully deleted.
+    """
+    logger = logging.getLogger(__name__)
+
+    # Before ES 7.10, deleting API keys by ID had to be done individually.
```

Comment:
We could actually get around this by deleting all API keys associated with the `basic_auth_user` that was supplied as a client option. We'd call `invalidate_api_key` with a `username` instead of `ids`/`id`. The API for this has existed since the introduction of API keys in 7.0, so we'd need to version-specific logic.

However, Rally would delete all of the user's API keys, not just the API keys that Rally created. If there are pre-existing API keys, this could be a problem. Benchmarking a cluster with any kind of pre-existing state is generally a bad idea, but I wouldn't want Rally to do something like invalidate a bunch of API keys just because a user made a mistake. So, if we want to go this route, we could first double-check that there are no existing API keys and fail with a helpful message if any are found, in addition to being clear about this in the docs.

It's a little funny that this component of the PR is perhaps the trickiest bit, which got me thinking about potential simplifications. What do you think @pquentin?

Reply:
I think the current code is fine, I mostly commented on details. Rally has a tradition of not messing with the cluster it's running on (https://esrally.readthedocs.io/en/stable/faq.html#will-rally-destroy-my-existing-indices), so let's keep it that way.

Reply:
Agreed, I prefer this approach, too.

Reply:
So this function is now super complex and hard to follow, which made me rethink this question. I still don't want to remove all keys and still want to take advantage of the bulk request, so I've tried to understand how we could improve the situation.

I think we are mixing two concerns here:

 * handle both the 7.10- and 7.10+ cases (ironically the latter became much more complex)
 * handle retries

I think we could have a `issue_delete_keys_call(ids, version)` function (with a better name?) that would encapsulate the versioning logic. And then we could handle retries the same way in both cases.

The retry itself is the most complex part. I'm not sure why we think sleeping will help? Not even sure why we think retries would help on deletion failure actually. I would expect that unless we hit a TransportError, then the issue is with the request, so retrying would be futile? It's slightly ridiculous we're calling `raise_exception()` four times.

Anyway, that's a lot of typing just to say that well, this works and is well tested, so let's ship it!